---
title: "References"

execute:
  eval: false

engine: knitr

bibliography: references.bib
---

This workflow makes use of tools developed by many different people, so please cite these tools accordingly. To make this easier you find here a list of tool requirements and (if available) the associated publications.

## Used databases

-   Silva fasta sequences: SILVA_138.1_SSURef_NR99 downloaded from [here](https://www.arb-silva.de/fileadmin/silva_databases/current/Exports/SILVA_138.1_SSURef_NR99_tax_silva.fasta.gz) [@quast2013]
-   Silva link to ncbi taxonomy: taxmap_embl-ebi_ena_ssu_ref_nr99_138.1.txt downloaded from [here](https://www.arb-silva.de/fileadmin/silva_databases/current/Exports/taxonomy/ncbi/taxmap_embl-ebi_ena_ssu_ref_nr99_138.1.txt.gz)
-   Unite fasta sequences: sh_general_release_all_18.07.2023 [@nilsson2019]
-   NCBI taxonomy dump downloaded on 19102023

## Used software

The version numbers are the exact versions used to develop this workflow.

-   Snakemake v7.32.4 [@m√∂lder2021]
-   Python v3.6.15
    -   biopython v1.79
    -   tabulate v0.8.10
    -   pandas v1.1.5
    -   matplotlib v3.3.4
    -   seaborn 0.11.2
-   R v4.2
    -   tidyverse v1.3.2 [@wickham2019]
    -   phyloseq v1.42.0 [@mcmurdie2013]
    -   microbiome v1.20.0 [@lahti2012]
-   NanoPack v1.1.0 [@decoster2023], which comes with the following tools used in this workflow:
    -   NanoStat v1.6.0
    -   Chopper v0.6.0
-   Pistis v0.3.3 [github_link](https://github.com/mbhall88/pistis)
-   Porechop v0.2.4 [github_link](<https://github.com/rrwick/Porechop>)
-   ITSx v1.1.3 [@bengtsson-palme2013]
-   Minimap2 v2.24 [@li2018]
-   Kraken2 v2.1.3 [@wood2019]


## Reference list

::: {#refs}
:::